fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexKeyValueToPartialRecordUtils.java

```diff
@@ -0,0 +1,182 @@
+            this.parent = parent;
+            this.descriptor = fieldDescriptor.getMessageType();
+            this.fieldDescriptor = fieldDescriptor;
+            this.builder = builder;
+        }
+
+        @Override
+        public Descriptors.Descriptor getDescriptor() {
+            return descriptor;
+        }
+
+        @Override
+        public Iterable<PartialRecordBuildSource> getChildren(@Nonnull FieldKeyExpression parentExpression) {
+            final String parentField = parentExpression.getFieldName();
+            final Descriptors.FieldDescriptor fieldDescriptor = descriptor.findFieldByName(parentField);
+            return Collections.singletonList(new PartialRecordBuildSource(this, fieldDescriptor, builder.newBuilderForField(fieldDescriptor)));
+        }
+
+        @Override
+        public Iterable<Object> getValues(@Nonnull FieldKeyExpression fieldExpression) {
+            return Collections.singletonList(fieldExpression.getFieldName());
+        }
+
+        public Message buildMessage(@Nonnull Object value, String field) {
+            return buildMessage(value, descriptor.findFieldByName(field));
+        }
+
+        private Message buildMessage(@Nonnull Object value, Descriptors.FieldDescriptor subFieldDescriptor) {
+            if (subFieldDescriptor.isRepeated()) {
+                if (builder.getRepeatedFieldCount(subFieldDescriptor) > 0) {
+                    Message.Builder subBuilder = builder.newBuilderForField(subFieldDescriptor);
+                    subBuilder
+                            .mergeFrom((Message) builder.getRepeatedField(subFieldDescriptor, 0))
+                            .mergeFrom((Message) value);
+                    builder.setRepeatedField(subFieldDescriptor, 0, subBuilder.build());
+                } else {
+                    builder.addRepeatedField(subFieldDescriptor, value);
+                }
+            } else {
+                builder.setField(subFieldDescriptor, value);
+            }
+
+            if (parent == null) {
+                return builder.build();
+            } else {
+                return parent.buildMessage(builder.build(), this.fieldDescriptor);
+            }
+        }
+
+    }
+
+    /**
+     * The copier to populate the lucene auto complete suggestion as a value for the field where it is indexed from.
+     * So the suggestion can be returned as a {@link com.apple.foundationdb.record.provider.foundationdb.FDBQueriedRecord} for query.
+     */
+    public static class LuceneAutoCompleteCopier implements IndexKeyValueToPartialRecord.Copier {
+
+        @Override
+        public void copy(@Nonnull Descriptors.Descriptor recordDescriptor, @Nonnull Message.Builder recordBuilder,
+                         @Nonnull IndexEntry kv) {
+            Tuple keyTuple = kv.getKey();
+            if (keyTuple.size() < 2) {
```

Comment:
There is a key and a value tuple in the indexEntry, isn't that supposed to be the value and field representation?